File: backend/worker-service/src/main/java/worker_service/worker_service/controller/WorkerController.java
package worker_service.worker_service.controller;

import lombok.RequiredArgsConstructor;
import org.springframework.core.io.FileSystemResource;
import org.springframework.core.io.Resource;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import worker_service.worker_service.entity.AuditLog;
import worker_service.worker_service.entity.StatementJob;
import worker_service.worker_service.service.AuditService;
import worker_service.worker_service.service.StatementService;

import java.io.File;
import java.util.List;

@RestController
@RequestMapping("/worker")
@RequiredArgsConstructor
public class WorkerController {

    private final StatementService statementService;
    private final AuditService auditService;

    @PostMapping("/statements/request")
    public ResponseEntity<StatementJob> requestStatement(@RequestParam Long userId) {
        return ResponseEntity.ok(statementService.requestStatement(userId));
    }

    @GetMapping("/statements/{jobId}")
    public ResponseEntity<StatementJob> getStatementStatus(@PathVariable Long jobId) {
        return ResponseEntity.ok(statementService.getJob(jobId));
    }

    @GetMapping("/statements/user/{userId}")
    public ResponseEntity<List<StatementJob>> getStatementsByUser(@PathVariable Long userId) {
        return ResponseEntity.ok(statementService.getJobsByUser(userId));
    }

    @GetMapping("/statements/{jobId}/download")
    public ResponseEntity<Resource> downloadStatement(@PathVariable Long jobId) {
        StatementJob job = statementService.getJob(jobId);
        if (!"COMPLETED".equalsIgnoreCase(job.getStatus()) || job.getPdfPath() == null) {
            return ResponseEntity.badRequest().build();
        }

        File file = new File(job.getPdfPath());
        if (!file.exists()) {
            return ResponseEntity.notFound().build();
        }

        Resource resource = new FileSystemResource(file);
        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"" + file.getName() + "\"")
                .contentType(MediaType.APPLICATION_PDF)
                .body(resource);
    }

    @GetMapping("/audit/user/{userId}")
    public ResponseEntity<List<AuditLog>> getAuditLogsByUser(@PathVariable Long userId) {
        return ResponseEntity.ok(auditService.getAuditLogsForUser(userId));
    }
}

File: backend-nestjs/worker-service/src/adapters/inbound/rest/worker.controller.ts
import {
  Controller,
  Get,
  Post,
  Param,
  Query,
  Body,
  Inject,
  Res,
  ParseIntPipe,
  BadRequestException,
  NotFoundException,
} from '@nestjs/common';
import { ApiTags, ApiOperation, ApiResponse } from '@nestjs/swagger';
import { Response } from 'express';
import * as fs from 'fs';
import * as path from 'path';
import {
  WORKER_SERVICE_PORT,
  WorkerServicePort,
} from '../../../domain/ports/worker-service.port';

@ApiTags('Worker')
@Controller('worker')
export class WorkerController {
  constructor(
    @Inject(WORKER_SERVICE_PORT)
    private readonly workerService: WorkerServicePort,
  ) {}

  @Get('health')
  @ApiOperation({ summary: 'Healthcheck para Ingress' })
  getHealth() {
    return { status: 'OK', service: 'worker-service', timestamp: new Date().toISOString() };
  }

  @Post('statements/request')
  @ApiOperation({ summary: 'Solicitar generación de extracto bancario en PDF' })
  @ApiResponse({ status: 201, description: 'Trabajo de extracto registrado' })
  async requestStatement(
    @Body() body?: any,
    @Query('userId') queryUserId?: string,
  ) {
    const rawId = body?.userId ?? (queryUserId ? parseInt(queryUserId, 10) : undefined);
    if (!rawId || isNaN(Number(rawId))) {
      throw new BadRequestException('userId es requerido en el body o query param');
    }
    return this.workerService.requestStatement(Number(rawId));
  }

  @Get('statements/:jobId')
  @ApiOperation({ summary: 'Obtener estado de un trabajo de extracto' })
  @ApiResponse({ status: 200, description: 'Estado del trabajo de extracto' })
  async getStatementStatus(@Param('jobId') jobId: string) {
    const numJobId = Number(jobId);
    if (isNaN(numJobId)) {
      throw new BadRequestException('jobId debe ser un número');
    }
    return this.workerService.getJob(numJobId);
  }

  @Get('statements/user/:userId')
  @ApiOperation({ summary: 'Obtener lista de trabajos de extractos de un usuario' })
  @ApiResponse({ status: 200, description: 'Lista de trabajos del usuario' })
  async getStatementsByUser(@Param('userId', ParseIntPipe) userId: number) {
    return this.workerService.getJobsByUser(userId);
  }

  @Get('statements/:jobId/download')
  @ApiOperation({ summary: 'Descargar archivo PDF del extracto bancario' })
  @ApiResponse({ status: 200, description: 'Archivo PDF del extracto' })
  async downloadStatement(@Param('jobId') jobId: string, @Res() res: Response) {
    const numJobId = Number(jobId);
    if (isNaN(numJobId)) {
      throw new BadRequestException('jobId debe ser un número');
    }
    const job = await this.workerService.getJob(numJobId);

    if (job.status !== 'COMPLETED' || !job.pdfPath) {
      throw new BadRequestException('El extracto aún no está listo o falló su generación');
    }

    const file = path.resolve(job.pdfPath);
    if (!fs.existsSync(file)) {
      throw new NotFoundException('El archivo PDF no fue encontrado en el servidor');
    }

    res.setHeader('Content-Type', 'application/pdf');
    res.setHeader('Content-Disposition', `attachment; filename="statement_job_${numJobId}.pdf"`);

    const filestream = fs.createReadStream(file);
    filestream.pipe(res);
  }

  @Get('audit/user/:userId')
  @ApiOperation({ summary: 'Obtener registros de auditoría de un usuario' })
  @ApiResponse({ status: 200, description: 'Lista de registros de auditoría' })
  async getAuditLogsByUser(@Param('userId', ParseIntPipe) userId: number) {
    return this.workerService.getAuditLogsForUser(userId);
  }
}
